internal/k8s/reconciler/manager.go

```diff
@@ -219,18 +225,18 @@ func (m *GatewayReconcileManager) upsertRoute(ctx context.Context, r Route, id s
 	managed, err := m.deleteUnmanagedRoute(ctx, route, id)
 	if err != nil {
 		return err
-	}
-	if !managed {
+	} else if !managed {
 		return nil
 	}
 
 	// Calling validate outside of the upsert process allows us to re-resolve any
 	// external references and set the statuses accordingly. Since we actually
 	// have other object updates triggering reconciliation loops, this is necessary
 	// prior to dirty-checking on upsert.
-	if err := route.validate(ctx); err != nil {
+	if err = m.routeValidator.Validate(ctx, route.RouteState, route.Route); err != nil {
```

Comment:
Is the plan to move the state initialization into the `Validate` call like (I believe) the previous refactor form had? Just wondering since the separation of the two main fields on a route/gateway seems a bit odd here.

Something more akin to:

```
route := &v1beta1.HTTPRoute{}
state, err := routeValidator.Validate(ctx, route)
...
store.Upsert(ctx, NewK8sRoute(route, state))
...
```

seems like it might be a bit of a cleaner abstraction -- the initial state coming from the validation of the original Kubernetes object and then the composing of the two into our "stateful route" right before upsert.

Reply:
It felt weird to me to have `.Validate()` spit out the state for a route it but thinking on it more, I believe it might just be the function name that bothers me. `Validate()` feels like it should answer "is this thing valid? true or false", not build a state object for the route. I'll spend some time this afternoon pondering, and maybe it will click or I can think up a name that resonates more.

Reply:
Yeah, I'm fine with a change in the name -- the main thing in my mind though is that the majority (with some exceptions since it normalizes the remote refs, though resolving those are a side effect of making sure the refs are actually valid) of what the validators do _is_ handle validation. The problem is with the fact that in the Kubernetes world giving a response to say, "this thing is invalid" is inherently a stateful operation because it's tied to the Status field in the object that needs to be updated.

So, don't mind if you change the name, but at least in the Kubernetes world, my mind doesn't separate out validation from what otherwise are stateful operations.